util/range.py

```diff
@@ -1,7 +1,10 @@
+from typing import Optional
+
+
 class Range:
-    def __init__(self, start=None, end=None):
-        self.start = start
-        self.end = end
+    def __init__(self, start: Optional[int] = None, end: Optional[int] = None):
+        self.start: Optional[int] = start
```

Comment:
these type annotations are redundant but why not

Reply:
I never know which ones I should annotate. Dunno if I should annotate the parameters or the members.

I guess just annotating the parameters should be enough (?)

Reply:
Yeah just the parameters is probably better here :)

Reply:
Using `dataclasses` would avoid this issue, btw, as well as making the code shorter. Applies as long as the classes don't do anything fancy in their constructors, which ime is true most of the time. `Symbol` could also be shortened that way